python/treelite/sklearn/exporter.py

```diff
@@ -0,0 +1,233 @@
+    Export as scikit-learn RandomForest or GradientBoosting.
+
+    Note
+    ----
+    Currently only random forests can be exported as scikit-learn model objects.
+    Support for gradient boosted trees and other kinds of tree models will be
+    added in the future.
+
+    Parameters
+    ----------
+    model : :py:class:`Model`
+        Treelite mobel to export
+
+    Returns
+    -------
+    sklearn_model : object of type \
+                    :py:class:`~sklearn.ensemble.RandomForestRegressor` / \
+                    :py:class:`~sklearn.ensemble.RandomForestClassifier` / \
+                    :py:class:`~sklearn.ensemble.GradientBoostingRegressor` / \
+                    :py:class:`~sklearn.ensemble.GradientBoostingClassifier`
+        Scikit-learn model
+    """
+    # pylint: disable=too-many-locals
+    try:
+        from sklearn import __version__ as sklearn_version
+        from sklearn.ensemble import RandomForestRegressor
+    except ImportError as e:
+        raise TreeliteError("This function requires scikit-learn package") from e
+
+    header_accessor = model.get_header_accessor()
+    average_tree_output = (
+        _ensure_scalar_int(header_accessor.get_field("average_tree_output")) == 1
+    )
+    n_features = _ensure_scalar_int(header_accessor.get_field("num_feature"))
+    n_trees = _ensure_scalar_int(header_accessor.get_field("num_tree"))
+    n_targets = _ensure_scalar_int(header_accessor.get_field("num_target"))
+    n_classes = _ensure_numpy(header_accessor.get_field("num_class"))
+    leaf_vector_shape = _ensure_numpy(header_accessor.get_field("leaf_vector_shape"))
+    target_id = _ensure_numpy(header_accessor.get_field("target_id"))
+    class_id = _ensure_numpy(header_accessor.get_field("class_id"))
+    tree_depths = model.get_tree_depth()
+
+    # Heuristics to ensure that the model can be represented as scikit-learn random forest
+    # 1. average_tree_output must be True
+    # 2. n_classes[i] must be identical for all targets
+    # 3. Each leaf must yield an output of shape (n_targets, n_classes)
+    # 4. target_id[i] must be either 0 or -1
+    # 5. class_id[i] must be either 0 or -1
+    def raise_not_rf_error():
+        raise NotImplementedError(
+            "Currently only random forests can be exported as scikit-learn model objects. "
+            "Gradient boosting models and other kinds of decision tree models are not yet "
+            "supported."
+        )
+
+    if not average_tree_output:
+        raise_not_rf_error()
+    if not np.all(n_classes == n_classes[0]):
+        raise_not_rf_error()
+    if not np.array_equal(leaf_vector_shape, [n_targets, n_classes.max()]):
+        raise_not_rf_error()
+    if not np.all((target_id == 0) | (target_id == -1)):
+        raise_not_rf_error()
+    if not np.all((class_id == 0) | (class_id == -1)):
+        raise_not_rf_error()
+
+    # Heuristics for determining whether the model is a regressor or a classifier
+    if n_classes[0] == 1:
```

Comment:
:( that we have to guess this and can't inspect `model` to get a definitive answer

Reply:
It turns out we can query the model type directly: `header_accessor.get_field("task_type")`

Reply:
Cool! Naively that seems like "the right thing to do" (tm) then no?

Reply:
Changed it to query the model type directly.